### PR TITLE
PLANET-7445 Migrate Split Two Columns block existing content

### PR DESCRIPTION
**Pull request by sagarsdeshmukh**
Add a migration script which transform Split two columns block into core columns block

Ref. https://jira.greenpeace.org/browse/PLANET-7445

### Testing:

On local dev env, run npx wp-env run cli wp p4-run-activator to run the migration script.
[Ref.](https://app.gitbook.com/o/-LMm4Q4AuKcwl38JYrxF/s/-M15KrJzoMvhbv4NcO9o/tech/data-migrations)

**Output of post deploy script execution on test instance -**
https://app.circleci.com/pipelines/github/greenpeace/planet4-test-titan/801/workflows/24380812-ed7a-4703-a282-406b071778cb/jobs/2939?invite=true#step-108-7176_16

**Comparison of posts before and after migration -**
[without migration](https://www-dev.greenpeace.org/test-mars/testing-all-blocks-evergreen-template/#this-is-a-columns-block-3)

[after migration](https://www-dev.greenpeace.org/test-titan/testing-all-blocks-evergreen-template/)

### CODE REVIEW

**Comment by planet-4** [comment]
### Test instance is ready :rocket:

:new_moon: [titan](https://www-dev.greenpeace.org/test-titan) | [admin](https://www-dev.greenpeace.org/test-titan/wp-admin/) | [blocks report](https://www-dev.greenpeace.org/test-titan/wp-admin/?page=plugin_blocks_report) | [CircleCI](https://app.circleci.com/pipelines/github/greenpeace/planet4-test-titan) | [composer-local.json](https://github.com/greenpeace/planet4-test-titan/blob/main/composer-local.json)

:watch: 2024.09.05 06:16:10

**Review by mardelnet — COMMENTED** [review]
Should we get rid of the mentions of the `split-two-columns`  on files `exporter-helper.php` and `src/Importer.php`?

File: src/Migrations/M034MigrateSplit2ColumnBlock.php [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+    // phpcs:enable SlevomatCodingStandard.Functions.UnusedParameter
+
+    /**
+     * Transform a block attrs into columns block.
+     *
+     * @param array $block - A block data array.
+     * @return array - The transformed block.
+     * phpcs:disable Generic.Files.LineLength.MaxExceeded
+     */
+    private static function get_transform_block(array $block): array
```
Comment (mardelnet) [inline src/Migrations/M034MigrateSplit2ColumnBlock.php:95]: @sagarsdeshmukh 
This is just a personal preference, so please ignore this message if you disagree.
IMO, using WordPress Block Code makes the code harder to read, debug, and maintain.
The same result can be achieved with PHP arrays (check it below).
I know, the extension of the code increases, but by using arrays we can take advantage of reusable functions or values pulled from the constants file:

<details>
  <summary>Click to expand</summary>

```php
$block = array(
                    'blockName' => 'core/columns',
                    'attrs' => array(),
                    'innerBlocks' => array(
                        array(
                            'blockName' => 'core/column',
                            'attrs' => array(
                                'verticalAlignment' => 'center',
                            ),
                            'innerBlocks' => array(
                                array(
                                    'blockName' => 'core/media-text',
                                    'attrs' => array(
                                        'mediaId' => 350,
                                        'mediaLink' => 'http://www.planet4.test/get-informed/people/tavolo-community-school-in-papua-new-guinea/',
                                        'mediaType' => 'image',
                                    ),
                                    'innerBlocks' => array(
                                        array(
                                            'blockName' => 'core/heading',
                                            'attrs' => array(
                                                'level' => 3,
                                            ),
                                            'innerBlocks' => array(),
                                            'innerHTML' => '<h3 class="wp-block-heading">People</h3>',
                                            'innerContent' => array(
                                                '<h3 class="wp-block-heading">People</h3>',
                                            ),
                                        ),
                                        array(
                                            'blockName' => 'core/paragraph',
                                            'attrs' => array(),
                                            'innerBlocks' => array(),
                                            'innerHTML' => '<p></p>',
                                            'innerContent' => array('<p></p>'),
                                        ),
                                        array(
                                            'blockName' => 'core/paragraph',
                                            'attrs' => array(),
                                            'innerBlocks' => array(),
                                            'innerHTML' => '<p><a href="http://www.planet4.test/get-informed/people/">Learn more about this issue</a></p>',
                                            'innerContent' => array('<p><a href="http://www.planet4.test/get-informed/people/">Learn more about this issue</a></p>'),
                                        ),
                                    ),
                                    'innerHTML' => '<div class="wp-block-media-text is-stacked-on-mobile"><figure class="wp-block-media-text__media"><img src="http://www.planet4.test/wp-content/uploads/2018/10/46697a18-people-6aa59e55-gp04o3g_pressmedia.jpg" alt="" class="wp-image-350 size-full"/></figure><div class="wp-block-media-text__content"></div></div>',
                                    'innerContent' => array(
                                        '<div class="wp-block-media-text is-stacked-on-mobile"><figure class="wp-block-media-text__media"><img src="http://www.planet4.test/wp-content/uploads/2018/10/46697a18-people-6aa59e55-gp04o3g_pressmedia.jpg" alt="" class="wp-image-350 size-full"/></figure><div class="wp-block-media-text__content">',
                                        '',
                                        '',
                                        '',
                                        '',
                                        '</div></div>',
                                    ),
                                ),
                            ),
                            'innerHTML' => '<div class="wp-block-column is-vertically-aligned-center"></div>',
                            'innerContent' => array(
                                '<div class="wp-block-column is-vertically-aligned-center">',
                                '',
                                '</div>',
                            ),
                        ),
                        array(
                            'blockName' => 'core/column',
                            'attrs' => array(),
                            'innerBlocks' => array(
                                array(
                                    'blockName' => 'core/media-text',
                                    'attrs' => array(
                                        'mediaId' => 347,
                                        'mediaLink' => 'http://www.planet4.test/consumption-f7fb64f4-gp0str6il_pressmedia/',
                                        'mediaType' => 'image',
                                    ),
                                    'innerBlocks' => array(
                                        array(
                                            'blockName' => 'core/heading',
                                            'attrs' => array(
                                                'level' => 3,
                                            ),
                                            'innerBlocks' => array(),
                                            'innerHTML' => '<h3 class="wp-block-heading">#Consumption</h3>',
                                            'innerContent' => array('<h3 class="wp-block-heading">#Consumption</h3>'),
                                        ),
                                        array(
                                            'blockName' => 'core/paragraph',
                                            'attrs' => array(),
                                            'innerBlocks' => array(),
                                            'innerHTML' => '<p>Buy. Use. Throw away. Buy more. Fast food, fast fashion, single use&hellip;</p>',
                                            'innerContent' => array('<p>Buy. Use. Throw away. Buy more. Fast food, fast fashion, single use&hellip;</p>'),
                                        ),
                                        array(
                                            'blockName' => 'core/buttons',
                                            'attrs' => array(),
                                            'innerBlocks' => array(
                                                array(
                                                    'blockName' => 'core/button',
                                                    'attrs' => array(
                                                        'className' => 'is-style-cta',
                                                    ),
                                                    'innerBlocks' => array(),
                                                    'innerHTML' => '<div class="wp-block-button is-style-cta"><a class="wp-block-button__link wp-element-button" href="http://www.planet4.test/tag/consumption/">Get involved</a></div>',
                                                    'innerContent' => array('<div class="wp-block-button is-style-cta"><a class="wp-block-button__link wp-element-button" href="http://www.planet4.test/tag/consumption/">Get involved</a></div>'),
                                                ),
                                            ),
                                            'innerHTML' => '<div class="wp-block-buttons"></div>',
                                            'innerContent' => array(
                                                '<div class="wp-block-buttons">',
                                                '',
                                                '</div>',
                                            ),
                                        ),
                                    ),
                                    'innerHTML' => '<div class="wp-block-media-text is-stacked-on-mobile"><figure class="wp-block-media-text__media"><img src="http://www.planet4.test/wp-content/uploads/2018/10/d9f1b57c-consumption-f7fb64f4-gp0str6il_pressmedia.jpg" alt="" class="wp-image-347 size-full"/></figure><div class="wp-block-media-text__content"></div></div>',
                                    'innerContent' => array(
                                        '<div class="wp-block-media-text is-stacked-on-mobile"><figure class="wp-block-media-text__media"><img src="http://www.planet4.test/wp-content/uploads/2018/10/d9f1b57c-consumption-f7fb64f4-gp0str6il_pressmedia.jpg" alt="" class="wp-image-347 size-full"/></figure><div class="wp-block-media-text__content">',
                                        '',
                                        '',
                                        '',
                                        '',
                                        '</div></div>',
                                    ),
                                ),
                            ),
                            'innerHTML' => '<div class="wp-block-column"></div>',
                            'innerContent' => array(
                                '<div class="wp-block-column">',
                                '',
                                '</div>',
                            ),
                        ),
                    ),
                    'innerHTML' => '<div class="wp-block-columns"></div>',
                    'innerContent' => array(
                        '<div class="wp-block-columns">',
                        '',
                        '',
                        '',
                        '',
                        '</div>',
                    ),
                );
                
            }
```

</details>

**Comment by sagarsdeshmukh** [comment]
Hi @mardelnet , Thank you for your feedback. I've updated the PR. please have a look.

File: src/Migrations/M034MigrateSplit2ColumnBlock.php [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+            echo "Split 2 Columns block migration in progress...\n"; // phpcs:ignore
+
+            foreach ($posts as $post) {
+                if (empty($post->post_content)) {
+                    continue;
+                }
+
+                $current_post_id = $post->ID; // Store the current post ID
+
+                echo 'Parsing post ', $post->ID, "\n"; // phpcs:ignore
```
Comment (mleray) [inline src/Migrations/M034MigrateSplit2ColumnBlock.php:47]: Could use `$$current_post_id` here

File: src/Migrations/M034MigrateSplit2ColumnBlock.php [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+                //$result = false;
```
Comment (mleray) [inline src/Migrations/M034MigrateSplit2ColumnBlock.php:48]: Does this comment need to be kept?

File: src/Migrations/M034MigrateSplit2ColumnBlock.php [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+                // Unset the reference to prevent potential issues.
+                unset($block);
+
+                // Serialize the blocks content.
+                $new_content = serialize_blocks($blocks);
+
+                if ($post->post_content !== $new_content) {
+                    $post_update = array(
+                        'ID' => $post->ID,
```
Comment (mleray) [inline src/Migrations/M034MigrateSplit2ColumnBlock.php:73]: Same comment as earlier, could use `$current_post_id`

File: src/Migrations/M034MigrateSplit2ColumnBlock.php [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+                        'post_content' => $new_content,
+                    );
+
+                    // Update the post with the replaced blocks.
+                    $result = wp_update_post($post_update);
+                    if ($result === 0) {
+                        throw new \Exception("There was an error trying to update the post #" . $post->ID);
```
Comment (mleray) [inline src/Migrations/M034MigrateSplit2ColumnBlock.php:80]: Again here we could use `$current_post_id`

**Review by mardelnet — COMMENTED** [review]
@sagarsdeshmukh please do not forget to open a complementary PR to remove the block from the plugin so we avoid editors to keep using it after the migration.

**Review by mardelnet — COMMENTED** [review]
According to the mockup in the JIRA ticket, after the migration, the content should go full-width instead of inside a container. Is that possible?

File: src/Migrations/M034MigrateSplit2ColumnBlock.php [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+
+        return $blocks[0];
+    }
+
+    /**
+     * Get the split-two-columns block attrs.
+     *
+     * @param array $block - A parsed split-two-columns block.
+     * @return array - A block attrs array.
+     * phpcs:disable Generic.Files.LineLength.MaxExceeded
```
Comment (mardelnet) [inline src/Migrations/M034MigrateSplit2ColumnBlock.php:156]: Do we need to disable this rule here?
It is already disabled on line 101.

File: src/Migrations/M034MigrateSplit2ColumnBlock.php [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+                    continue;
+                }
+
+                $current_post_id = $post->ID; // Store the current post ID
+
+                echo 'Parsing post ', $post->ID, "\n"; // phpcs:ignore
+                //$result = false;
+
+                // Go through the post blocks to find the planet4-blocks/split-two-columns one.
+                $blocks = $parser->parse($post->post_content);
```
Comment (mardelnet) [inline src/Migrations/M034MigrateSplit2ColumnBlock.php:51]: We could add an additional control here:

```suggestion
                $blocks = $parser->parse($post->post_content);
                if (!is_array($blocks)) {
                    throw new \Exception("Invalid block structure for post #" . $current_post_id);
                }
```

File: src/Migrations/M034MigrateSplit2ColumnBlock.php [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+                    );
+
+                    // Update the post with the replaced blocks.
+                    $result = wp_update_post($post_update);
+                    if ($result === 0) {
+                        throw new \Exception("There was an error trying to update the post #" . $post->ID);
+                    }
+                }
+
+                echo $result
```
Comment (mardelnet) [inline src/Migrations/M034MigrateSplit2ColumnBlock.php:84]: Is this alright?
If the condition `$post->post_content !== $new_content` is false, then `$result` would never be declared before this line.

**Comment by comzeradd** [comment]
> According to the mockup in the JIRA ticket, after the migration, the content should go full-width instead of inside a container. Is that possible?

@sagarsdeshmukh can you share an example of the current implementation with the design team and check if they are happy with the result. I guess we could use `.block-wide` here but since it's just a columns block it may look weird to have it full width.

**Comment by mardelnet** [comment]
@sagarsdeshmukh 
I will continue the migration on this PR: https://github.com/greenpeace/planet4-master-theme/pull/2433
Do you mind if I close this PR?

**Comment by mardelnet** [comment]
Closing this PR in favor of https://github.com/greenpeace/planet4-master-theme/pull/2433

**Pull request closed without merging**